### ISSUE TITLE
Dangling filter doesn't work properly when used on images prune

### ISSUE THREAD

**Issue by rdlf0** [moby/moby#33591]
**Description**
I'm trying to remove all dangling images through the `Docker API`, using `cURL`, but the **dangling** filter doesn't seem to work as expected.

**Steps to reproduce the issue:**
When I execute `docker images` through the cli, I get this list:
```
REPOSITORY                                  TAG                 IMAGE ID            CREATED             SIZE
my.repository.com/project/image          master              31a8bb572943        About an hour ago   523 MB
my.repository.com/project/image          <none>              b1bf9e0bcee0        About an hour ago   523 MB
my.repository.com/project/image          <none>              ae59986ac910        4 hours ago         523 MB
my.repository.com/anotherproject/image   web                 6f5664d4e09c        2 days ago          45.7 MB
my.repository.com/anotherproject/image   <none>              05dde9468fad        2 days ago          42 MB
my.repository.com/anotherproject/image   <none>              5c51a672d505        2 days ago          42 MB
```
I have 4 running containers - 3 of them are using the first image `(my.repository.com/project/image:master)` and one is using `my.repository.com/anotherproject/image:web`. I removed all stopped containers, so there is no difference between the results of `docker ps` and `docker ps -a`, therefore the `<none>` images should be unused and untagged.
According to [the documentation](https://docs.docker.com/engine/api/v1.29/#operation/ImagePrune) the **dangling** filter should be set to **true**, to match these **unused and untagged** images and prune them, but after few hours of tests it turned out that the images are matched and pruned only if the filter is set to **false**, which, according to the documentation, should work only with unused images. So, it turns out that my `<none>` images are tagged somehow.
Here is the command I run, which successfully pruned my `<none>` images:
```
curl --unix-socket /var/run/docker.sock -G -X POST "http:/v1.29/images/prune" --data-urlencode 'filters={"dangling":{"false": true}}'
```
If this can help, the `<none>` images are result of pulling their new versions and recreation of `docker-compose` services.

**Describe the results you received:**
The `<none>` images aren't matched when using `dangling=true` filter. They are matched only with `dangling=false` filter.

**Describe the results you expected:**
The `<none>` images have to be matched and pruned when using `dangling=true` filter.

**Additional information you deem important (e.g. issue happens only occasionally):**
Happens all the time.

**Output of `docker version`:**

```
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:05:44 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:05:44 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 17.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: false
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.6.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.64 GiB
Name: mydockermachine
ID: IT5W:6NWN:PZ3F:PHPB:E4YE:V7VL:WJEL:PDGO:RZHU:LYKR:SJIH:QLKL
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical personal machine

**Comment by thaJeztah** [comment]
I haven't checked the API docs, but this is what the CLI does;

Prune all images (only dangling)

    docker image prune

Performs this API call;

    Calling POST /v1.26/images/prune?filters={"dangling":{"true":true}}

Prune all images (not just dangling)

    docker image prune -a

Performs this API call;

    Calling POST /v1.26/images/prune?filters={"dangling":{"false":true}}



Testing on Docker 17.03, it looks like dangling images are not removed (i.e. not using `-a`);

```bash
docker pull busybox
docker pull hello-world
docker tag hello-world busybox:latest

docker images


REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
busybox             <none>              c75bebcdd211        3 weeks ago         1.11 MB
busybox             latest              48b5124b2768        4 months ago        1.84 kB
hello-world         latest              48b5124b2768        4 months ago        1.84 kB
```


```bash
$ docker image prune -f
Total reclaimed space: 0 B

$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
busybox             <none>              c75bebcdd211        3 weeks ago         1.11 MB
busybox             latest              48b5124b2768        4 months ago        1.84 kB
hello-world         latest              48b5124b2768        4 months ago        1.84 kB
```


However, doing the same on Docker 17.05

```bash
$ docker image prune -f
Deleted Images:
untagged: busybox@sha256:c79345819a6882c31b41bc771d9a94fc52872fa651b36771fbe0c8461d7ee558
deleted: sha256:c75bebcdd211f41b3a460c7bf82970ed6c75acaab9cd4c9a4e125b03ca113798
deleted: sha256:4ac76077f2c741c856a2419dfdb0804b18e48d2e1a9ce9c6a3f0605a2078caba

Total reclaimed space: 1.106MB
```

And removing "all" images;

```bash
$ docker image prune -a -f
Deleted Images:
untagged: busybox:latest
untagged: hello-world:latest
untagged: hello-world@sha256:c5515758d4c5e1e838e9cd307f6c6a0d620b5e07e6f927b07d05f6d12a1ac8d7
deleted: sha256:48b5124b2768d2b917edcb640435044a97967015485e812545546cbed5cf0233
deleted: sha256:98c944e98de8d35097100ff70a31083ec57704be0991a92c51700465e4544d08

Total reclaimed space: 1.84kB
```


Let me try to find the PR that fixed this

**Comment by thaJeztah** [comment]
Looks like this is a duplicate of https://github.com/moby/moby/issues/30307, https://github.com/moby/moby/issues/30583,  and was fixed in https://github.com/moby/moby/pull/30330 (which is in docker 17.04 and up)

I'll close this issue, because this is fixed already, and part of 17.04, 17.05 (edge channel), and will be part of 17.06 (stable channel), but feel free to continue the discussion
